### ISSUE TITLE
whitelist is not working

### ISSUE THREAD

**Issue by shomodj** [tomzo/gocd-yaml-config-plugin#61]
invert_filter is not set to true when using whitelist, there is no mention it in the code

**Comment by tomzo** [comment]
You can see how to use whitelist [under here](https://github.com/tomzo/gocd-yaml-config-plugin#git). You either specify `whitelist` or `blacklist`. 

What do you mean by ?
> invert_filter is not set to true when using whitelist

How did you configure pipeline material? What do you expect from whitelist.

**Comment by shomodj** [comment]
gocd version: 17.12.0
material type: hg

```  
image_nexus:
    group: images
    materials:
      xxx:
        hg: ssh://hg@bitbucket.org/xxx
        whitelist:
          - images/nexus/**/*
```

curl to get pipeline settings
```
  "materials": [
    {
      "type": "hg",
      "attributes": {
        "url": "ssh://hg@bitbucket.org/xxx",
        "destination": null,
        "filter": {
          "ignore": [
            "images/nexus/**/*"
          ]
        },
        "invert_filter": false,
        "name": "xxx",
        "auto_update": true
      }
    }
  ],
```

now lets try blacklist
```
  image_nexus:
    group: images
    materials:
      xxx:
        hg: ssh://hg@bitbucket.org/xxx
        blacklist:
          - images/nexus/**/*
```

curl
```
  "materials": [
    {
      "type": "hg",
      "attributes": {
        "url": "ssh://hg@bitbucket.org/xxx",
        "destination": null,
        "filter": {
          "ignore": [
            "images/nexus/**/*"
          ]
        },
        "invert_filter": false,
        "name": "xxx",
        "auto_update": true
      }
    }
  ],
```

on whitelist, invert_filter should be true, I have tried the same setup manually using UI and it works as expected, I have a bunch of docker images and they all rebuild even if I change only one, whitelist should prevent this

**Comment by shomodj** [comment]
Hey, were you able to confirm this?

**Comment by tomzo** [comment]
I started looking at this, but then ran into problems into getting intellij working on gocd gradle project due to some silly errors.
I need to debug both server and plugin to get why this is not working. I am sure whitelist works fine on git, because it's part of tests. No clue why it would not work on hg. But that API response is certainly unexpected.

**Comment by shomodj** [comment]
Thanks

**Comment by arvindsv** [comment]
@tomzo You know this code best, but [this line](https://github.com/gocd/gocd/blob/86f5ac705ba6ea7cc2bf0ec8d96a2b87eb41514d/server/src/main/java/com/thoughtworks/go/config/ConfigConverter.java#L367) seems to be missed [here](https://github.com/gocd/gocd/blob/86f5ac705ba6ea7cc2bf0ec8d96a2b87eb41514d/server/src/main/java/com/thoughtworks/go/config/ConfigConverter.java#L371). As you can see [in the next line](https://github.com/gocd/gocd/blob/86f5ac705ba6ea7cc2bf0ec8d96a2b87eb41514d/server/src/main/java/com/thoughtworks/go/config/ConfigConverter.java#L372), `invertFilter` is always set to `false`.

**Comment by tomzo** [comment]
@arvindsv thanks, that was it.
The fix was easy, but I've spent about 2 hours trying to figure out why I cannot import project to intellij. Apparently older version `2017.1` does not work with gradle 4.x, at least for GoCD.
The PR is testing https://github.com/gocd/gocd/pull/4191 , I'll merge once it passes.

**Comment by shomodj** [comment]
Any chance of a release to address this? 0.6.1 ?

**Comment by tomzo** [comment]
@shomodj the bug is fixed in the core. You'll need to upgrade server to version 18.1

**Comment by shomodj** [comment]
Cool, thanks

**Comment by tomzo** [comment]
Fixed in server 18.1+
